### PR TITLE
ENH: tracking quality module

### LINKED ISSUE #92: Data quality assessments
The data quality assessments by time (tracking quality per hour/day/week per user) are already implemented and need to be integrated into trackintel. The current idea is to put it into the `analysis` module and output a quality df. The user can then decide whether to plot or save it.

Question: how do we measure data quality spatially?

### PR DESCRIPTION
**Pull request by hong2223**
closes #92

corresponding tests and docs are added.

### CODE REVIEW

File: tests/analysis/test_tracking_quality.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
+import sys
+import os
+import glob
+import pytest
+
+sys.path.append(os.path.join(os.getcwd(), "trackintel"))
```
Comment (hong2223) [inline tests/analysis/test_tracking_quality.py:6]: I think this is not needed, will be removed in the next commit

File: trackintel/analysis/tracking_quality.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    """
+    df = source.copy()
+    if granularity == "day":
+        change_flag = df["started_at"].dt.date != df["finished_at"].dt.date
+    elif granularity == "hour":
+        change_flag = df["started_at"].dt.hour != df["finished_at"].dt.hour
+    i = 0
+
+    while change_flag.sum() > 0:
+
```
Comment (henrymartin1) [inline trackintel/analysis/tracking_quality.py:115]: Maybe add a one-liner how the algorithm works in general:
Something like: 
Iteratively split away one day/hour from multi day/hour entries until no entry spans over multiple days/hours

File: tests/analysis/test_tracking_quality.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        """Test if _split_overlaps() function can split records that span several days."""
+        stps_tpls = testdata_stps_tpls_geolife_long
+
+        # some of the records spans several day
+        multi_day_records = stps_tpls["finished_at"].dt.day - stps_tpls["started_at"].dt.day
+        assert (multi_day_records > 0).any()
+
+        # split the records according to day
+        stps_tpls.reset_index(inplace=True)
+        spliited = ti.analysis.tracking_quality._split_overlaps(stps_tpls, granularity="day")
```
Comment (henrymartin1) [inline tests/analysis/test_tracking_quality.py:90]: consider refactoring "spliited" to "splitted"

**Review by henrymartin1 — CHANGES_REQUESTED** [review]
Hey, 
great addition! Finally some content for the analysis part ;-)
I have added some comments that should be addressed most of them are optional though

File: trackintel/analysis/tracking_quality.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    quality: DataFrame
+        A per-user temporal tracking quality dataframe.
+
+    Note
+    ----
+    The temporal tracking quality is the time proportion of tracked period with the possible
+    time extent. The possible time extent of the different granularities are different: "all"
+    considers the time between the latest "finished_at" and the earliest "started_at", whereas
+    "day" considers the whole day (86399 sec) and "hour" considers the whole hour (3599 sec).
+    """
```
[the remark this answers is not in the document]
Comment (hong2223) [reply trackintel/analysis/tracking_quality.py:30]: I thought one day has only 86399 sec. Apparently wrong...

File: trackintel/analysis/tracking_quality.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+        raise AttributeError(f"granularity unknown. We only support ['all', 'day', 'hour']. You passed {granularity}")
+    return quality
+
+
+def _get_tracking_quality_user(df, granularity="all"):
+    """Tracking quality per-user per-granularity.
+
+    Parameters
+    ----------
+    df : GeoDataFrame (as trackintel datamodels)
```
[the remark this answers is not in the document]
Comment (hong2223) [reply trackintel/analysis/tracking_quality.py:63]: df has to have the columns "finished_at" and "started_at" and sometimes also "day". I guess the safest is to say it should be a trackintel model. 
This function is only called inside, so I think it does not matter so much anyway.

File: trackintel/analysis/tracking_quality.py [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+    elif granularity == "hour":
+        # total seconds of a day times number of different days
+        extent = (60 * 60 - 1) * len(df["day"].unique())
+        print(df[["started_at", "finished_at"]])
+        print(extent)
+    return pd.Series([tracked_duration / extent], index=["quality"])
+
+
+def _split_overlaps(source, granularity="day"):
+    """Split input df that have a duration of several days or hours.
```
[the remark this answers is not in the document]
Comment (hong2223) [reply trackintel/analysis/tracking_quality.py:91]: Maybe this function does the job for all the users. 
What it does is basically to duplicate the entries that span several day/hour and change the time info. user_id is not just duplicated

File: trackintel/analysis/tracking_quality.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (henrymartin1) [reply trackintel/analysis/tracking_quality.py:63]: 👍

File: trackintel/analysis/tracking_quality.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (henrymartin1) [reply trackintel/analysis/tracking_quality.py:91]: Yes, I think you are right.

File: trackintel/analysis/tracking_quality.py [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+    pandas.Series
+        A pandas.Series object containing the tracking quality
+    """
+    tracked_duration = (df["finished_at"] - df["started_at"]).dt.total_seconds().sum()
+    if granularity == "all":
+        # the whole tracking period
+        extent = (df["finished_at"].max() - df["started_at"].min()).total_seconds()
+    elif granularity == "day":
+        # total seconds of a day
+        extent = 60 * 60 * 24 - 1
```
[the remark this answers is not in the document]
Comment (hong2223) [reply trackintel/analysis/tracking_quality.py:81]: grouping by day is unique -> multiple days would not exist in one df; theoretical upper limit is just `sec_of_day`
grouping by hour is not unique -> it is possible that entries from different days are grouped into the same hour df; the tracked_duration thus has a theoretical upper limit of `sec_of_hour * unique_days`

**Comment by hong2223** [comment]
The comments shall be fixed and additional tests are added.

**Comment by henrymartin1** [comment]
Nice contribution. This is really super useful, I think I can directly use it 👍

**Pull request merged**